File: go/07-estado-contexto/mini-simulador-ventana.go
// Mini-proyecto: El simulador de ventana de contexto (Go).
//
// Uso:
//
//	go run mini-simulador-ventana.go
//	go run mini-simulador-ventana.go -ventana 4096
//	go run mini-simulador-ventana.go -turnos 30 -tecnica clearing
//	go run mini-simulador-ventana.go -tecnica todos

// Cómo ejecutar: make go FILE=go/07-estado-contexto/mini-simulador-ventana.go

package main

import (
	"flag"
	"fmt"
	"math/rand"
	"strings"
)

// Snapshot precios mayo 2026 — verificar en docs del proveedor
const precioSonnetInput = 3.00 // USD / millón tokens
const ventanaDefault = 8_192

// ── tipos ─────────────────────────────────────────────────────────────────────

type Mensaje struct {
	role    string
	content string
	bloques []Bloque // nil para mensajes de texto plano
}

type Bloque struct {
	tipo       string // "tool_use" | "tool_result"
	id         string
	contenido  string
	clareado   bool
}

func (m *Mensaje) texto() string {
	if len(m.bloques) > 0 {
		var parts []string
		for _, b := range m.bloques {
			parts = append(parts, b.contenido)
		}
		return strings.Join(parts, " ")
	}
	return m.content
}

func estimarTokens(texto string) int {
	n := len([]rune(texto)) / 4
	if n < 1 {
		return 1
	}
	return n
}

func tokensMensaje(m *Mensaje) int {
	return estimarTokens(m.texto()) + 4
}

func tokensHistorial(mensajes []*Mensaje) int {
	total := 0
	for _, m := range mensajes {
		total += tokensMensaje(m)
	}
	return total
}

// ── generador ─────────────────────────────────────────────────────────────────

var palabras = []string{
	"agente", "contexto", "herramienta", "respuesta", "análisis",
	"código", "función", "resultado", "iteración", "plan",
	"decisión", "estado", "memoria", "búsqueda", "resumen",
}

func lorem(nTokens int, rng *rand.Rand) string {
	var parts []string
	for len(strings.Join(parts, " "))/4 < nTokens {
		parts = append(parts, palabras[rng.Intn(len(palabras))])
	}
	return strings.Join(parts, " ")
}

var tiposPeso = []struct {
	tipo string
	peso int
}{
	{"user_simple", 30},
	{"user_largo", 120},
	{"assistant_texto", 80},
	{"tool_call", 40},
	{"tool_result_corto", 60},
	{"tool_result_largo", 400},
}

func elegirPonderado(rng *rand.Rand) string {
	total := 0
	for _, tp := range tiposPeso {
		total += tp.peso
	}
	r := rng.Intn(total)
	for _, tp := range tiposPeso {
		r -= tp.peso
		if r < 0 {
			return tp.tipo
		}
	}
	return tiposPeso[len(tiposPeso)-1].tipo
}

func generarTurno(tipo string, rng *rand.Rand) *Mensaje {
	switch tipo {
	case "user_simple":
		return &Mensaje{role: "user", content: lorem(30, rng)}
	case "user_largo":
		return &Mensaje{role: "user", content: lorem(120, rng)}
	case "assistant_texto":
		return &Mensaje{role: "assistant", content: lorem(80, rng)}
	case "tool_call":
		id := fmt.Sprintf("t%d", rng.Intn(9000)+1000)
		return &Mensaje{
			role: "assistant",
			bloques: []Bloque{{tipo: "tool_use", id: id, contenido: lorem(10, rng)}},
		}
	case "tool_result_corto":
		id := fmt.Sprintf("t%d", rng.Intn(9000)+1000)
		return &Mensaje{
			role: "user",
			bloques: []Bloque{{tipo: "tool_result", id: id, contenido: lorem(60, rng)}},
		}
	case "tool_result_largo":
		id := fmt.Sprintf("t%d", rng.Intn(9000)+1000)
		return &Mensaje{
			role: "user",
			bloques: []Bloque{{tipo: "tool_result", id: id, contenido: lorem(400, rng)}},
		}
	}
	return &Mensaje{role: "user", content: lorem(30, rng)}
}

func generarHistorial(nTurnos int) []*Mensaje {
	rng := rand.New(rand.NewSource(42))
	msgs := make([]*Mensaje, nTurnos)
	for i := range msgs {
		msgs[i] = generarTurno(elegirPonderado(rng), rng)
	}
	return msgs
}

// ── técnicas ──────────────────────────────────────────────────────────────────

func esToolResult(m *Mensaje) bool {
	for _, b := range m.bloques {
		if b.tipo == "tool_result" {
			return true
		}
	}
	return false
}

func clonarMensaje(m *Mensaje) *Mensaje {
	nuevo := &Mensaje{role: m.role, content: m.content}
	if len(m.bloques) > 0 {
		nuevo.bloques = make([]Bloque, len(m.bloques))
		copy(nuevo.bloques, m.bloques)
	}
	return nuevo
}

func clonarHistorial(msgs []*Mensaje) []*Mensaje {
	clon := make([]*Mensaje, len(msgs))
	for i, m := range msgs {
		clon[i] = clonarMensaje(m)
	}
	return clon
}

func aplicarClearing(msgs []*Mensaje) []*Mensaje {
	resultado := make([]*Mensaje, len(msgs))
	for i, m := range msgs {
		if esToolResult(m) {
			nuevo := clonarMensaje(m)
			for j := range nuevo.bloques {
				if nuevo.bloques[j].tipo == "tool_result" {
					nuevo.bloques[j].contenido = "[cleared]"
				}
			}
			resultado[i] = nuevo
		} else {
			resultado[i] = m
		}
	}
	return resultado
}

func aplicarHeadTail(msgs []*Mensaje, maxTokens int) []*Mensaje {
	if tokensHistorial(msgs) <= maxTokens {
		return msgs
	}
	var head, tail []*Mensaje
	presupuesto := maxTokens
	for _, m := range msgs {
		tok := tokensMensaje(m)
		if presupuesto >= tok {
			head = append(head, m)
			presupuesto -= tok
		} else {
			break
		}
	}
	for i := len(msgs) - 1; i >= len(head); i-- {
		tok := tokensMensaje(msgs[i])
		if presupuesto >= tok {
			tail = append([]*Mensaje{msgs[i]}, tail...)
			presupuesto -= tok
		} else {
			break
		}
	}
	omitidos := len(msgs) - len(head) - len(tail)
	sep := &Mensaje{role: "user", content: fmt.Sprintf("[... %d mensajes omitidos ...]", omitidos)}
	return append(append(head, sep), tail...)
}

func aplicarSumarizacion(msgs []*Mensaje, maxTokens int, rng *rand.Rand) []*Mensaje {
	if tokensHistorial(msgs) <= maxTokens {
		return msgs
	}
	nConservar := len(msgs) / 3
	if nConservar < 2 {
		nConservar = 2
	}
	aResumir := msgs[:len(msgs)-nConservar]
	recientes := msgs[len(msgs)-nConservar:]
	tokResumidos := tokensHistorial(aResumir)
	resumenTok := tokResumidos / 5
	if resumenTok < 20 {
		resumenTok = 20
	}
	resumen := &Mensaje{
		role: "user",
		content: fmt.Sprintf("[RESUMEN de %d mensajes / ~%d tokens → %d tokens]: %s",
			len(aResumir), tokResumidos, resumenTok, lorem(resumenTok, rng)),
	}
	return append([]*Mensaje{resumen}, recientes...)
}

// ── simulación ────────────────────────────────────────────────────────────────

type Resultado struct {
	tecnica         string
	tokensEnviados  int
	compactaciones  int
	tokensAhorrados int
	desbordamientos int
	tokensFinal     int
	costoUSD        float64
}

func simular(original []*Mensaje, tecnica string, ventana int) Resultado {
	umbral := int(float64(ventana) * 0.85)
	rng := rand.New(rand.NewSource(99))

	var compactaciones, tokensEnviados, tokensAhorrados, desbordamientos int
	var historial []*Mensaje

	for _, msg := range original {
		historial = append(historial, msg)
		tokActual := tokensHistorial(historial)

		if tokActual > umbral {
			tokAntes := tokActual
			switch tecnica {
			case "clearing":
				historial = aplicarClearing(historial)
			case "head_tail":
				historial = aplicarHeadTail(historial, umbral)
			case "sumarizacion":
				historial = aplicarSumarizacion(historial, umbral, rng)
			}
			compactaciones++
			diff := tokAntes - tokensHistorial(historial)
			if diff > 0 {
				tokensAhorrados += diff
			}
		}

		tokFinal := tokensHistorial(historial)
		tokensEnviados += tokFinal
		if tokFinal > ventana {
			desbordamientos++
		}
	}

	return Resultado{
		tecnica:         tecnica,
		tokensEnviados:  tokensEnviados,
		compactaciones:  compactaciones,
		tokensAhorrados: tokensAhorrados,
		desbordamientos: desbordamientos,
		tokensFinal:     tokensHistorial(historial),
		costoUSD:        float64(tokensEnviados) * precioSonnetInput / 1_000_000,
	}
}

// ── presentación ──────────────────────────────────────────────────────────────

func barra(valor, maximo, ancho int) string {
	if maximo == 0 {
		return strings.Repeat("░", ancho)
	}
	lleno := valor * ancho / maximo
	return strings.Repeat("█", lleno) + strings.Repeat("░", ancho-lleno)
}

func imprimirResultados(resultados []Resultado, nTurnos, ventana int) {
	fmt.Printf("\n%s\n", strings.Repeat("=", 66))
	fmt.Println("  SIMULADOR DE VENTANA DE CONTEXTO")
	fmt.Printf("  %d turnos  |  ventana %d tokens  |  precios sonnet mayo 2026\n", nTurnos, ventana)
	fmt.Printf("%s\n", strings.Repeat("=", 66))

	maxTok := 1
	maxCost := 0.0001
	for _, r := range resultados {
		if r.tokensEnviados > maxTok {
			maxTok = r.tokensEnviados
		}
		if r.costoUSD > maxCost {
			maxCost = r.costoUSD
		}
	}

	fmt.Printf("\n%-16s %11s %9s %11s %9s %8s\n",
		"Técnica", "Tokens env.", "Compact.", "Ahorr. tok", "Desbord.", "USD")
	fmt.Println(strings.Repeat("-", 66))
	for _, r := range resultados {
		fmt.Printf("%-16s %11d %9d %11d %9d $%7.4f\n",
			r.tecnica, r.tokensEnviados, r.compactaciones,
			r.tokensAhorrados, r.desbordamientos, r.costoUSD)
	}

	fmt.Printf("\n%s\n", strings.Repeat("─", 66))
	fmt.Println("  Tokens enviados (barra relativa al máximo)")
	fmt.Println(strings.Repeat("─", 66))
	for _, r := range resultados {
		fmt.Printf("  %-14s %s  %d\n", r.tecnica, barra(r.tokensEnviados, maxTok, 30), r.tokensEnviados)
	}

	fmt.Printf("\n%s\n", strings.Repeat("─", 66))
	fmt.Println("  Costo USD (barra relativa al máximo)")
	fmt.Println(strings.Repeat("─", 66))
	var base *Resultado
	for i := range resultados {
		if resultados[i].tecnica == "ninguna" {
			base = &resultados[i]
			break
		}
	}
	for _, r := range resultados {
		ahorroStr := ""
		if base != nil && r.tecnica != "ninguna" {
			ahorro := (1 - r.costoUSD/base.costoUSD) * 100
			ahorroStr = fmt.Sprintf("  (%+.1f%% vs sin compactación)", ahorro)
		}
		fmt.Printf("  %-14s %s  $%.4f%s\n", r.tecnica, barra(int(r.costoUSD*1e6), int(maxCost*1e6), 30), r.costoUSD, ahorroStr)
	}

	fmt.Println("\n[Estimación ±10% — conteo exacto con tiktoken (Python)]")
	fmt.Println("[Snapshot precios mayo 2026 — verificar en docs del proveedor]")
}

// ── main ──────────────────────────────────────────────────────────────────────

func main() {
	ventana := flag.Int("ventana", ventanaDefault, "Tokens de ventana")
	turnos := flag.Int("turnos", 40, "Número de turnos a simular")
	tecnicaFlag := flag.String("tecnica", "todos", "Técnica: clearing|head_tail|sumarizacion|ninguna|todos")
	flag.Parse()

	historial := generarHistorial(*turnos)
	tokensBruto := tokensHistorial(historial)

	fmt.Printf("\n[Historial generado: %d turnos, %d tokens bruto]\n", *turnos, tokensBruto)
	fmt.Printf("[Ventana configurada: %d tokens  |  umbral: %d]\n", *ventana, int(float64(*ventana)*0.85))

	var tecnicas []string
	if *tecnicaFlag == "todos" {
		tecnicas = []string{"ninguna", "clearing", "head_tail", "sumarizacion"}
	} else {
		tecnicas = []string{*tecnicaFlag}
	}

	resultados := make([]Resultado, len(tecnicas))
	for i, t := range tecnicas {
		resultados[i] = simular(historial, t, *ventana)
	}
	imprimirResultados(resultados, *turnos, *ventana)
}

File: typescript/07-estado-contexto/mini-simulador-ventana.ts
/**
 * Mini-proyecto: El simulador de ventana de contexto (TypeScript).
 *
 * Uso:
 *   npx ts-node mini-simulador-ventana.ts
 *   npx ts-node mini-simulador-ventana.ts --ventana 4096
 *   npx ts-node mini-simulador-ventana.ts --turnos 30 --tecnica clearing
 *   npx ts-node mini-simulador-ventana.ts --tecnica todos
 */

// Snapshot precios mayo 2026 — verificar en docs del proveedor

// Cómo ejecutar: make ts SCRIPT=typescript/07-estado-contexto/mini-simulador-ventana.ts

const PRECIO_SONNET_INPUT = 3.00; // USD / millón tokens
const VENTANA_DEFAULT = 8_192;

type Mensaje = {
  role: string;
  content: string | Array<Record<string, unknown>>;
};

function estimarTokens(texto: string): number {
  return Math.max(1, Math.floor(texto.length / 4));
}

function tokensMensaje(msg: Mensaje): number {
  let texto = "";
  if (typeof msg.content === "string") {
    texto = msg.content;
  } else if (Array.isArray(msg.content)) {
    texto = msg.content
      .map((b) => (typeof b === "object" ? JSON.stringify(b) : String(b)))
      .join(" ");
  }
  return estimarTokens(texto) + 4;
}

function tokensHistorial(mensajes: Mensaje[]): number {
  return mensajes.reduce((acc, m) => acc + tokensMensaje(m), 0);
}

// ── generador de conversación ficticia ───────────────────────────────────────

const TIPOS_TURNO: [string, number][] = [
  ["user_simple", 30],
  ["user_largo", 120],
  ["assistant_texto", 80],
  ["tool_call", 40],
  ["tool_result_corto", 60],
  ["tool_result_largo", 400],
];

const PALABRAS = ["agente", "contexto", "herramienta", "respuesta", "análisis",
  "código", "función", "resultado", "iteración", "plan",
  "decisión", "estado", "memoria", "búsqueda", "resumen"];

let seed = 42;
function rand(): number {
  seed = (seed * 1664525 + 1013904223) & 0xffffffff;
  return (seed >>> 0) / 0xffffffff;
}
function randInt(min: number, max: number): number {
  return Math.floor(rand() * (max - min + 1)) + min;
}
function randChoice<T>(arr: T[]): T {
  return arr[Math.floor(rand() * arr.length)];
}
function weightedChoice(tipos: string[], pesos: number[]): string {
  const total = pesos.reduce((a, b) => a + b, 0);
  let r = rand() * total;
  for (let i = 0; i < tipos.length; i++) {
    r -= pesos[i];
    if (r <= 0) return tipos[i];
  }
  return tipos[tipos.length - 1];
}

function lorem(nTokens: number): string {
  const palabras: string[] = [];
  while (palabras.join(" ").length / 4 < nTokens) {
    palabras.push(randChoice(PALABRAS));
  }
  return palabras.join(" ");
}

function generarTurno(tipo: string): Mensaje {
  if (tipo === "user_simple") return { role: "user", content: lorem(30) };
  if (tipo === "user_largo") return { role: "user", content: lorem(120) };
  if (tipo === "assistant_texto") return { role: "assistant", content: lorem(80) };
  if (tipo === "tool_call") {
    return {
      role: "assistant",
      content: [{ type: "tool_use", id: `t${randInt(1000, 9999)}`,
        name: "search_docs", input: { query: lorem(10) } }],
    };
  }
  if (tipo === "tool_result_corto") {
    return {
      role: "user",
      content: [{ type: "tool_result", tool_use_id: `t${randInt(1000, 9999)}`,
        content: lorem(60) }],
    };
  }
  if (tipo === "tool_result_largo") {
    return {
      role: "user",
      content: [{ type: "tool_result", tool_use_id: `t${randInt(1000, 9999)}`,
        content: lorem(400) }],
    };
  }
  return { role: "user", content: lorem(30) };
}

function generarHistorial(nTurnos: number): Mensaje[] {
  seed = 42;
  const tipos = TIPOS_TURNO.map(([t]) => t);
  const pesos = TIPOS_TURNO.map(([, p]) => p);
  return Array.from({ length: nTurnos }, () => generarTurno(weightedChoice(tipos, pesos)));
}

// ── técnicas de compactación ──────────────────────────────────────────────────

function esToolResult(msg: Mensaje): boolean {
  if (!Array.isArray(msg.content)) return false;
  return msg.content.some((b) => typeof b === "object" && (b as Record<string,unknown>).type === "tool_result");
}

function clearing(mensajes: Mensaje[]): Mensaje[] {
  return mensajes.map((msg) => {
    if (!esToolResult(msg)) return msg;
    const nuevo = { ...msg, content: (msg.content as Array<Record<string,unknown>>).map((b) => {
      if (typeof b === "object" && b.type === "tool_result") {
        return { ...b, content: "[cleared]" };
      }
      return b;
    })};
    return nuevo;
  });
}

function headTail(mensajes: Mensaje[], maxTokens: number): Mensaje[] {
  if (tokensHistorial(mensajes) <= maxTokens) return mensajes;
  const head: Mensaje[] = [];
  const tail: Mensaje[] = [];
  let presupuesto = maxTokens;
  for (const msg of mensajes) {
    const tok = tokensMensaje(msg);
    if (presupuesto >= tok) { head.push(msg); presupuesto -= tok; }
    else break;
  }
  for (let i = mensajes.length - 1; i >= head.length; i--) {
    const tok = tokensMensaje(mensajes[i]);
    if (presupuesto >= tok) { tail.unshift(mensajes[i]); presupuesto -= tok; }
    else break;
  }
  const omitidos = mensajes.length - head.length - tail.length;
  const sep: Mensaje = { role: "user", content: `[... ${omitidos} mensajes omitidos ...]` };
  return [...head, sep, ...tail];
}

function sumarizacionSimulada(mensajes: Mensaje[], maxTokens: number): Mensaje[] {
  if (tokensHistorial(mensajes) <= maxTokens) return mensajes;
  const nConservar = Math.max(2, Math.floor(mensajes.length / 3));
  const aResumir = mensajes.slice(0, mensajes.length - nConservar);
  const recientes = mensajes.slice(mensajes.length - nConservar);
  const tokResumidos = tokensHistorial(aResumir);
  const resumenTok = Math.max(20, Math.floor(tokResumidos / 5));
  const resumen: Mensaje = {
    role: "user",
    content: `[RESUMEN de ${aResumir.length} mensajes / ~${tokResumidos} tokens → ${resumenTok} tokens]: ${lorem(resumenTok)}`,
  };
  return [resumen, ...recientes];
}

// ── simulación ────────────────────────────────────────────────────────────────

type Resultado = {
  tecnica: string;
  tokensEnviados: number;
  compactaciones: number;
  tokensAhorrados: number;
  desbordamientos: number;
  tokensFinal: number;
  costoUsd: number;
};

function simular(historialOriginal: Mensaje[], tecnica: string, ventana: number): Resultado {
  const umbral = Math.floor(ventana * 0.85);
  let historial: Mensaje[] = [];
  let compactaciones = 0;
  let tokensEnviados = 0;
  let tokensAhorrados = 0;
  let desbordamientos = 0;

  for (const msg of historialOriginal) {
    historial.push(msg);
    const tokActual = tokensHistorial(historial);

    if (tokActual > umbral) {
      const tokAntes = tokActual;
      if (tecnica === "clearing") historial = clearing(historial);
      else if (tecnica === "head_tail") historial = headTail(historial, umbral);
      else if (tecnica === "sumarizacion") historial = sumarizacionSimulada(historial, umbral);
      // "ninguna" — no hace nada
      compactaciones++;
      tokensAhorrados += Math.max(0, tokAntes - tokensHistorial(historial));
    }

    const tokFinal = tokensHistorial(historial);
    tokensEnviados += tokFinal;
    if (tokFinal > ventana) desbordamientos++;
  }

  return {
    tecnica,
    tokensEnviados,
    compactaciones,
    tokensAhorrados,
    desbordamientos,
    tokensFinal: tokensHistorial(historial),
    costoUsd: tokensEnviados * PRECIO_SONNET_INPUT / 1_000_000,
  };
}

// ── presentación ──────────────────────────────────────────────────────────────

function barra(valor: number, maximo: number, ancho = 30): string {
  const lleno = maximo ? Math.round((valor / maximo) * ancho) : 0;
  return "█".repeat(lleno) + "░".repeat(ancho - lleno);
}

function imprimirResultados(resultados: Resultado[], nTurnos: number, ventana: number): void {
  console.log(`\n${"=".repeat(66)}`);
  console.log(`  SIMULADOR DE VENTANA DE CONTEXTO`);
  console.log(`  ${nTurnos} turnos  |  ventana ${ventana.toLocaleString()} tokens  |  precios sonnet mayo 2026`);
  console.log("=".repeat(66));

  const maxTok = Math.max(...resultados.map((r) => r.tokensEnviados));
  const maxCost = Math.max(...resultados.map((r) => r.costoUsd));

  console.log(`\n${"Técnica".padEnd(16)} ${"Tokens env.".padStart(11)} ${"Compact.".padStart(9)} ${"Ahorr. tok".padStart(11)} ${"Desbord.".padStart(9)} ${"USD".padStart(8)}`);
  console.log("-".repeat(66));
  for (const r of resultados) {
    console.log(
      `${r.tecnica.padEnd(16)} ${r.tokensEnviados.toLocaleString().padStart(11)} ${r.compactaciones.toString().padStart(9)} ` +
      `${r.tokensAhorrados.toLocaleString().padStart(11)} ${r.desbordamientos.toString().padStart(9)} $${r.costoUsd.toFixed(4).padStart(7)}`
    );
  }

  console.log(`\n${"─".repeat(66)}`);
  console.log("  Tokens enviados (barra relativa al máximo)");
  console.log("─".repeat(66));
  for (const r of resultados) {
    console.log(`  ${r.tecnica.padEnd(14)} ${barra(r.tokensEnviados, maxTok)}  ${r.tokensEnviados.toLocaleString()}`);
  }

  console.log(`\n${"─".repeat(66)}`);
  console.log("  Costo USD (barra relativa al máximo)");
  console.log("─".repeat(66));
  const base = resultados.find((r) => r.tecnica === "ninguna");
  for (const r of resultados) {
    let ahorroStr = "";
    if (base && r.tecnica !== "ninguna") {
      const ahorro = (1 - r.costoUsd / base.costoUsd) * 100;
      ahorroStr = `  (${ahorro >= 0 ? "+" : ""}${ahorro.toFixed(1)}% vs sin compactación)`;
    }
    console.log(`  ${r.tecnica.padEnd(14)} ${barra(r.costoUsd, maxCost)}  $${r.costoUsd.toFixed(4)}${ahorroStr}`);
  }

  console.log(`\n[Estimación ±10% — conteo exacto con tiktoken (Python)]`);
  console.log(`[Snapshot precios mayo 2026 — verificar en docs del proveedor]`);
}

// ── main ──────────────────────────────────────────────────────────────────────

function main(): void {
  const args = process.argv.slice(2);
  const getArg = (flag: string, def: string) => {
    const i = args.indexOf(flag);
    return i >= 0 && args[i + 1] ? args[i + 1] : def;
  };

  const ventana = parseInt(getArg("--ventana", String(VENTANA_DEFAULT)));
  const turnos = parseInt(getArg("--turnos", "40"));
  const tecnicaArg = getArg("--tecnica", "todos");

  const historial = generarHistorial(turnos);
  const tokensBruto = tokensHistorial(historial);

  console.log(`\n[Historial generado: ${turnos} turnos, ${tokensBruto.toLocaleString()} tokens bruto]`);
  console.log(`[Ventana configurada: ${ventana.toLocaleString()} tokens  |  umbral: ${Math.floor(ventana * 0.85).toLocaleString()}]`);

  const tecnicas = tecnicaArg === "todos"
    ? ["ninguna", "clearing", "head_tail", "sumarizacion"]
    : [tecnicaArg];

  const resultados = tecnicas.map((t) => simular([...historial], t, ventana));
  imprimirResultados(resultados, turnos, ventana);
}

main();
